File: app/src/main/java/com/ilyap/yutaold/ui/dialogs/project/EditProjectDialog.java
package com.ilyap.yutaold.ui.dialogs.project;

import android.content.Context;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;
import androidx.fragment.app.Fragment;
import com.ilyap.yutaold.R;
import com.ilyap.yuta.domain.entity.Project;
import com.ilyap.yuta.domain.response.ProjectResponse;
import com.ilyap.yuta.domain.entity.Team;
import com.ilyap.yuta.domain.response.UpdateResponse;
import com.ilyap.yutaold.ui.fragments.ProjectsFragment;
import com.ilyap.yuta.utils.FileUtils;
import com.ilyap.yuta.domain.entity.ProjectStatus;
import lombok.SneakyThrows;

import java.io.InputStream;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

import static android.view.View.VISIBLE;

@SuppressWarnings("ConstantConditions")
public class EditProjectDialog extends CreateProjectDialog {
    private final int projectId;
    private Project project;
    private Spinner spinner;

    public EditProjectDialog(Context context, Fragment fragment, int projectId) {
        super(context, fragment);
        this.projectId = projectId;
    }

    @Override
    public void start() {
        super.start();
        setupViews();
        getProject(projectId);
        statusSpinnerInitialize();

        submitButton.setEnabled(false);
        submitButton.setOnClickListener(v -> {
            hideKeyboard(submitButton);
            editProject();
        });
    }

    private void setupViews() {
        submitButton.setText(R.string.save_button);
        ((TextView) dialog.findViewById(R.id.create_text)).setText(getContext().getString(R.string.edit_project));
        dialog.findViewById(R.id.status_container).setVisibility(VISIBLE);
    }

    private void getProject(int projectId) {
        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.getProject(projectId);
        viewModel.getResultLiveData().observe(fragment, result -> {
            if (!(result instanceof ProjectResponse)) return;
            project = ((ProjectResponse) result).getProject();
            setupProject();
        });
    }

    private void setupProject() {
        List<Team> teamList = new ArrayList<>();
        if (project.getTeam() != null) {
            radioGroup.check(RADIO_CREATE_WITH_TEAM);
            teamList.add(project.getTeam());
            addedTeamSearchAdapter.updateList(teamList);
        }

        projectName.setText(project.getName());
        projectDesc.setText(project.getDescription());
        fileName.setText(project.getTechnicalTaskName());
        spinner.setSelection(ProjectStatus.fromText(project.getStatus()).ordinal());
        deadlineField.setText(getUnformattedDate(project.getDeadline()));

        updateAddedTextVisibility();
    }

    @SneakyThrows
    private void editProject() {
        String name = getData(projectName);
        String description = getData(projectDesc);
        String deadline = getFormattedDate(getData(deadlineField));
        String status = spinner.getSelectedItem().toString();
        InputStream techTaskInputStream = techTaskUri != null ? getContext().getContentResolver().openInputStream(techTaskUri) : null;
        String filename = FileUtils.getFileName(fragment.requireContext(), techTaskUri);
        int teamId = getCurrentTeamId();

        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.editProject(projectId, name, description, deadline, filename, status, techTaskInputStream, teamId);
        viewModel.getResultLiveData().observe(fragment, result -> {
            if (!(result instanceof UpdateResponse)) return;
            ((ProjectsFragment) fragment).updateLists();
            dismiss();
        });
    }

    private String getUnformattedDate(String date) {
        DateTimeFormatter sourceDateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        LocalDate localDate = LocalDate.parse(date, sourceDateFormatter);

        return localDate.format(DateTimeFormatter.ofPattern("dd.MM.yyyy"));
    }

    private void statusSpinnerInitialize() {
        spinner = dialog.findViewById(R.id.status);
        ArrayAdapter<ProjectStatus> adapter = new ArrayAdapter<>(
                getContext(),
                android.R.layout.simple_spinner_item,
                ProjectStatus.values()
        );

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (project != null && position != ProjectStatus.fromText(project.getStatus()).ordinal()) {
                    updateSubmitButtonState();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
    }
}

File: app/src/main/java/com/yuta/projects/ui/dialog/EditProjectDialog.kt
package com.yuta.projects.ui.dialog

import android.view.View
import android.view.View.VISIBLE
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.viewModelScope
import com.yuta.app.R
import com.yuta.common.util.DateUtils
import com.yuta.common.util.FieldUtils.getData
import com.yuta.common.util.FileUtils
import com.yuta.common.util.KeyboardUtils
import com.yuta.domain.model.ProjectDto
import com.yuta.domain.model.ProjectStatus
import com.yuta.domain.model.Team
import com.yuta.projects.viewmodel.ProjectDialogsViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class EditProjectDialog(
    private val projectDto: ProjectDto,
    private val fragment: Fragment,
    private val onEditSuccessCallback: () -> Unit
) : CreateProjectDialog(fragment) {

    companion object {
        private val RADIO_CREATE_WITH_TEAM: Int = R.id.create_with_team
    }

    private val spinner: Spinner by lazy { dialog.findViewById(R.id.status) }
    private val title: TextView by lazy { dialog.findViewById(R.id.title) }
    private val statusContainer: View by lazy { dialog.findViewById(R.id.status_container) }

    private val projectViewModel: ProjectDialogsViewModel by fragment.viewModels()

    override fun start() {
        super.start()
        setupViews()
        initializeStatusSpinner()
        populateProjectDetails()

        submitButton.setOnClickListener {
            KeyboardUtils.hideKeyboard(fragment.requireActivity(), submitButton)
            editProject()
        }
    }

    private fun setupViews() {
        submitButton.apply {
            setText(R.string.save_button)
            isEnabled = false
        }
        title.setText(R.string.edit_project_text)
        statusContainer.visibility = VISIBLE
    }

    private fun populateProjectDetails() {
        val teamList = mutableListOf<Team>()

        projectViewModel.viewModelScope.launch {
            val project = projectViewModel.getById(projectDto.id).first()

            project.team?.let {
                radioGroup.check(RADIO_CREATE_WITH_TEAM)
                teamList.add(it)
                addedTeamSearchAdapter.refillList(teamList)
            }

            fileName.text = project.technicalTaskName
            deadlineField.text = DateUtils.formatToEuropean(project.deadline)
            projectName.setText(project.name)
            projectDesc.setText(project.description)
            spinner.setSelection(ProjectStatus.fromText(project.status).ordinal)

            updateAddedTextVisibility()
        }
    }

    private fun editProject() {
        projectViewModel.viewModelScope.launch {
            val result = projectViewModel.edit(
                id = projectDto.id,
                name = projectName.getData().toString(),
                description = projectDesc.getData().toString(),
                deadline = DateUtils.formatToIso(deadlineField.getData().toString()),
                teamId = projectViewModel.addedTeams.firstOrNull()?.id,
                status = spinner.selectedItem.toString(),
                filename = FileUtils.getFilename(fragment.requireContext(), techTaskUri),
                technicalTask = techTaskUri?.let { context.contentResolver.openInputStream(it) }
            ).first()

            if (result) {
                onEditSuccessCallback()
                dismiss()
            }
        }
    }

    private fun initializeStatusSpinner() {
        spinner.apply {
            adapter = ArrayAdapter(
                context,
                android.R.layout.simple_spinner_item,
                ProjectStatus.entries.toTypedArray()
            ).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }

            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    if (position != ProjectStatus.fromText(projectDto.status).ordinal) {
                        updateSubmitButtonState()
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
    }
}
